### ISSUE TITLE
ClassCastException when receiving SAML LogoutResponse

### ISSUE THREAD

**Issue by timforsyth** [pac4j/pac4j#1570]
Hi,

I'm using Pac4j with Play Framework 2.8 using the latest 4.0.0 release (play-pac4j and pac4j-saml-opensamlv3)

I'm seeing a ClassCastException when receiving a LogoutResponse from my IDP:

Caused by: java.lang.ClassCastException: org.opensaml.saml.saml2.core.impl.LogoutResponseImpl cannot be cast to org.opensaml.saml.saml2.core.Response
	at org.pac4j.saml.profile.impl.AbstractSAML2MessageReceiver.receiveMessage(AbstractSAML2MessageReceiver.java:61)
	at org.pac4j.saml.logout.impl.SAML2LogoutProfileHandler.receive(SAML2LogoutProfileHandler.java:36)
	at org.pac4j.saml.credentials.extractor.SAML2CredentialsExtractor.extract(SAML2CredentialsExtractor.java:59)
	at org.pac4j.core.client.BaseClient.retrieveCredentials(BaseClient.java:61)
	at org.pac4j.core.client.IndirectClient.getCredentials(IndirectClient.java:140)
	at org.pac4j.core.engine.DefaultCallbackLogic.perform(DefaultCallbackLogic.java:81)
	at org.pac4j.play.CallbackController.lambda$callback$0(CallbackController.java:54)
	at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1590)
	... 7 common frames omitted

It looks like this regression was caused by this merge:

https://github.com/pac4j/pac4j/commit/d1c10240eabd8a3343bfd05ab79a435e08cee141#diff-3276716761b33bc5ca73d61859d97119

I'm fairly new to this project, but looks like all received messages go though AbstractSML2MessageReceiver and the assumption from the merge is that all objects passing through here are Reponse objects, which in this case they are not.

**Comment by leleuj** [comment]
Yes, it seems so.

@mmoayyed If I remember correctly, you introduced that change. Do you have a fix in mind?

At worst, we can simply test if it's a `Response`  to do: `final AssertionConsumerService acsService = context.getSPAssertionConsumerService((Response) decodedCtx.getMessageContext().getMessage());` and otherwise: `final AssertionConsumerService acsService = context.getSPAssertionConsumerService();`

**Comment by mmoayyed** [comment]
I'll take a look.

**Comment by leleuj** [comment]
Thanks. It'll go in the 4.0.1 fix release I planned for the beginning of next month.

**Comment by mmoayyed** [comment]
@timforsyth we have a patch merged in. please test the latest snapshots and let us know what works and what doesn't :)

**Comment by leleuj** [comment]
@mmoayyed The patch was only for the `pac4j-saml` module. The `pac4j-saml-opensamlv3` module has been forgotten...

**Comment by mmoayyed** [comment]
Thanks for the catch. Will apply changes to that module too.

**Comment by leleuj** [comment]
Thanks

**Comment by timforsyth** [comment]
Hey, thanks for the update.

I haven't tried it yet but looking at the new commit, I think there is still an issue.

Although the correct super type is now being used, dispatching a call across to validate the ACS URL isn't going to work because the destination in the response is a logout endpoint.

Maybe let the subclasses of AbstractSAML2MessageReceiver opt in or out of the endpoint checks?

**Comment by leleuj** [comment]
Can you test to confirm the problem?

**Comment by timforsyth** [comment]
I manually patched in the changes to the pac4j-saml-opensamlv3 module (I had to adapt slightly due to the change in SAML2MessageContext from v3 to v4).

I can confirm that a logout response causes a SAMLException because an attempt is made to validate it as an ACS endpoint.

```
Caused by: org.pac4j.saml.exceptions.SAMLException: Assertion consumer service with destination http://web.sp.test:9000/callback?client_name=SAML2Client&logoutendpoint=true could not be found for spDescriptor org.opensaml.saml.saml2.metadata.impl.SPSSODescriptorImpl@79fda29e
	at org.pac4j.saml.context.SAML2MessageContext.getSPAssertionConsumerService(SAML2MessageContext.java:126)
	at org.pac4j.saml.profile.impl.AbstractSAML2MessageReceiver.receiveMessage(AbstractSAML2MessageReceiver.java:64)
	at org.pac4j.saml.logout.impl.SAML2LogoutProfileHandler.receive(SAML2LogoutProfileHandler.java:36)
	at org.pac4j.saml.credentials.extractor.SAML2CredentialsExtractor.extract(SAML2CredentialsExtractor.java:59)
	at org.pac4j.core.client.BaseClient.retrieveCredentials(BaseClient.java:61)
	at org.pac4j.core.client.IndirectClient.getCredentials(IndirectClient.java:140)
	at org.pac4j.core.engine.DefaultCallbackLogic.perform(DefaultCallbackLogic.java:81)
	at org.pac4j.play.CallbackController.lambda$callback$0(CallbackController.java:54)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1700)
	... 8 common frames omitted
```

**Comment by mmoayyed** [comment]
@timforsyth Would you mind putting together a test case that demonstrates the failure? I am a little short on time these days, so if you can duplicate the failure with a unit test, I can take a look and fix. Tried the test route myself, but it seems more involved than what I have time for at the moment.

**Comment by leleuj** [comment]
@mmoayyed If you don't have time, I will do some tests

@timforsyth Don't hesitate to submit a PR with a fix if you already have one in mind!

**Comment by mmoayyed** [comment]
Thanks; that would be helpful. I'll try to look into this some more by Friday. If you get to it quicker, sure.

**Comment by leleuj** [comment]
I have been able to reproduce the issue. It should be fixed by this simple commit: https://github.com/pac4j/pac4j/commit/ca250474deea8d23585b3574eaf8331c04f953b3

@timforsyth Can you confirm it's fixed with the latest v4.0.2-SNAPSHOT? Thx

**Comment by timforsyth** [comment]
There is a further issue with this fix, which I've fix in PR https://github.com/pac4j/pac4j/pull/1600

I've deferred the endpoint lookup to the subclasses of AbstractSAML2MessageReceiver, which the Logout implementation opts out.

The SAML2LogoutValidator grabs the logout URL directly from the SP descriptor so a future enhancement might be to unify the approach taken in the Authn validator by looking at the current context.

`Caused by: org.pac4j.saml.exceptions.SAMLException: Assertion consumer service with destination http://test.sp:9000/callback?client_name=SAML2Client&logoutendpoint=true could not be found for spDescriptor org.opensaml.saml.saml2.metadata.impl.SPSSODescriptorImpl@433f19c9 at org.pac4j.saml.context.SAML2MessageContext.getSPAssertionConsumerService(SAML2MessageContext.java:126) at org.pac4j.saml.profile.impl.AbstractSAML2MessageReceiver.receiveMessage(AbstractSAML2MessageReceiver.java:64) at org.pac4j.saml.logout.impl.SAML2LogoutProfileHandler.receive(SAML2LogoutProfileHandler.java:36) at org.pac4j.saml.credentials.extractor.SAML2CredentialsExtractor.extract(SAML2CredentialsExtractor.java:59) at org.pac4j.core.client.BaseClient.retrieveCredentials(BaseClient.java:66) at org.pac4j.core.client.IndirectClient.getCredentials(IndirectClient.java:143) at org.pac4j.core.engine.DefaultCallbackLogic.perform(DefaultCallbackLogic.java:85) at org.pac4j.play.CallbackController.lambda$callback$0(CallbackController.java:54) at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1700) ... 8 common frames omitted`
